### PR TITLE
feat(docker): quay api v1 for tags

### LINKED ISSUE #9176: Fetching tags from docker registry is limited to 20 pages
**What would you like Renovate to be able to do?**

At the moment pagination is limited to 20 pages:
https://github.com/renovatebot/renovate/blob/master/lib/datasource/docker/index.ts#L458-L464
this proves problematic in certain registries/repositories as this means that renovate might not fetch _all_ tags, but only a subset. Having the option to remove/set the value for this limit will help mitigating this problem.

**Did you already have any implementation ideas?**

As per the discussion in https://github.com/renovatebot/renovate/discussions/9159, possible solutions would be to add a configuration option to change the default page limit/maximum tags limit.

I can attempt to write it myself .

### PR DESCRIPTION
**Pull request by giokara-work**
## Changes:

Use quay.io/api/v1 API to get Docker tags as documented in [the docs](https://access.redhat.com/documentation/en-us/red_hat_quay/3/html/red_hat_quay_api_guide/appendix_a_red_hat_quay_application_programming_interface_api#get_api_v1_repository_repository_tag).

## Context:

Closes #9176. Based on unfinished draft PR https://github.com/renovatebot/renovate/pull/9362 and replaces https://github.com/renovatebot/renovate/pull/9813.
Tested in https://github.com/oqton/test-renovate-docker2.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/docker/index.spec.ts [hunk 1]
```diff
@@ -408,7 +407,8 @@ describe(getName(), () => {
           200,
           { tags },
           {
-            link: '<https://api.github.com/user/9287/repos?page=3&per_page=100>; rel="next", ',
+            link:
+              '<https://api.github.com/user/9287/repos?page=3&per_page=100>; rel="next", ',
```
Comment (viceice) [inline lib/datasource/docker/index.spec.ts:411]: This url looks wrong to me either. Shouldn't it point to the docker registry?

File: lib/datasource/docker/index.spec.ts [hunk 2]
```diff
@@ -447,6 +447,27 @@ describe(getName(), () => {
       expect(res.releases).toHaveLength(1);
       expect(httpMock.getTrace()).toMatchSnapshot();
     });
+    it('uses quay api', async () => {
+      const tags = [{ name: '1.0.0' }];
+      httpMock
+        .scope('https://quay.io')
+        .get('/v2/')
+        .reply(200, '', {})
+        .get('/api/v1/repository/node/tag/?limit=10000&page=1')
```
Comment (viceice) [inline lib/datasource/docker/index.spec.ts:456]: Did you tested it? Does 10000 work?

File: lib/datasource/docker/quay.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    const limit = 10000;
+
+    let page = 1;
+    let url = `${registry}/api/v1/repository/${repository}/tag/?limit=${limit}&page=${page}`;
+    const headers = await getAuthHeaders(registry, repository);
+    if (!headers) {
+      logger.debug('Failed to get authHeaders for getTags lookup');
+      return null;
+    }
```
Comment (viceice) [inline lib/datasource/docker/quay.ts:30]:
```suggestion
```
Remove this. Auth should be handled by http layer.

File: lib/datasource/docker/quay.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    do {
+      const res = await http.getJson<{
+        tags: { name: string }[];
+        has_additional: boolean;
+      }>(url, {
+        headers,
```
Comment (viceice) [inline lib/datasource/docker/quay.ts:36]:
```suggestion
```
See other comment.

File: lib/datasource/docker/quay.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      });
+      const pageTags = res.body.tags.map((tag) => tag.name);
+      url = res.body.has_additional
+        ? `${registry}/api/v1/repository/${repository}/tag/?limit=${limit}&page=${page}&onlyActiveTags=true`
```
Comment (viceice) [inline lib/datasource/docker/quay.ts:40]: This url looks different to the first. Please unify.

File: lib/datasource/docker/quay.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    return tags;
+  } catch (err) /* istanbul ignore next */ {
+    if (err instanceof ExternalHostError) {
+      throw err;
+    }
+    if (err.statusCode === 404 && !repository.includes('/')) {
+      logger.debug(
+        `Retrying Tags for ${registry}/${repository} using library/ prefix`
+      );
+      return getTagsQuayRegistry('library/' + repository);
```
Comment (viceice) [inline lib/datasource/docker/quay.ts:56]: Remove this fallback, docker won't do this either.

File: lib/datasource/docker/quay.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      url = res.body.has_additional
+        ? `${registry}/api/v1/repository/${repository}/tag/?limit=${limit}&page=${page}&onlyActiveTags=true`
+        : null;
+      tags = tags.concat(pageTags);
+      page += 1;
+    } while (url && page < 20);
+    const cacheMinutes = 30;
+    await packageCache.set(cacheNamespace, cacheKey, tags, cacheMinutes);
+    return tags;
+  } catch (err) /* istanbul ignore next */ {
```
Comment (viceice) [inline lib/datasource/docker/quay.ts:48]: Please add tests for this instead of ignore.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/docker/index.spec.ts (hunk 1 above)
Comment (giokara-oqton) [reply lib/datasource/docker/index.spec.ts]: This was an incorrect lint change - this has not been adjusted in this PR.

File: lib/datasource/docker/index.spec.ts (hunk 2 above)
Comment (giokara-oqton) [reply lib/datasource/docker/index.spec.ts]: Limit is 100 while testing, updated it.

**Review by monstro805 — APPROVED** [review]

**Review by monstro805 — DISMISSED** [review]

File: lib/datasource/docker/quay.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
+import { logger } from '../../logger';
+import { ExternalHostError } from '../../types/errors/external-host-error';
+import * as packageCache from '../../util/cache/package';
+import { http } from './common';
+
+export async function getTagsQuayRegistry(
+  repository: string
+): Promise<string[] | null> {
```
Comment (JamieMagee) [inline lib/datasource/docker/quay.ts:8]: I can't see an explicit `return null` below. Can this return signature be simplified?

Comment (viceice) [reply lib/datasource/docker/quay.ts:8]:
```suggestion
): Promise<string[]> {
```
Yeah, simplify

File: lib/datasource/docker/quay.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    const cacheNamespace = 'datasource-docker-tags';
+    const cacheKey = `${registry}:${repository}`;
+    const cachedResult = await packageCache.get<string[]>(
+      cacheNamespace,
+      cacheKey
+    );
+    // istanbul ignore if
+    if (cachedResult !== undefined) {
+      return cachedResult;
+    }
```
Comment (viceice) [inline lib/datasource/docker/quay.ts:21]: Move cache handling a level up to deduplicate code, you can rename the `getTags` to something like `getDockerApiTags` and add a new `getTags` which handles the quay test and the caching.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/docker/index.spec.ts [hunk 10]
```diff
@@ -447,6 +447,43 @@ describe(getName(), () => {
...
+          '/api/v1/repository/node/tag/?limit=100&page=1&onlyActiveTags=true'
+        )
+        .reply(200, { tags, has_additional: false })
+        .get('/v2/')
+        .reply(200, '', {})
+        .get('/v2/node/manifests/1.0.0')
+        .reply(200, '', {});
+      const config = {
+        datasource: id,
+        depName: 'node',
```
Comment (viceice) [inline lib/datasource/docker/index.spec.ts:464]: can we use a real dep name? as i know `node` isn't supported. maybe [`bitnami/redis`](https://quay.io/repository/bitnami/redis)  is a good sample.

File: lib/datasource/docker/__snapshots__/index.spec.ts.snap [hunk 11]
```diff
@@ -1175,3 +1175,40 @@ Array [
...
+  Object {
+    "headers": Object {
+      "accept": "application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.docker.distribution.manifest.v2+json",
+      "accept-encoding": "gzip, deflate, br",
+      "authorization": "Basic c29tZS11c2VybmFtZTpzb21lLXBhc3N3b3Jk",
+      "host": "quay.io",
+      "user-agent": "https://github.com/renovatebot/renovate",
+    },
+    "method": "GET",
+    "url": "https://quay.io/v2/bitnami/redis/manifests/1.0.0",
```
Comment (viceice) [inline lib/datasource/docker/__snapshots__/index.spec.ts.snap:1211]:
```suggestion
    "url": "https://quay.io/v2/bitnami/redis/manifests/5.0.12",
```

Let's use an existing tag.

File: lib/datasource/docker/index.spec.ts [hunk 12]
```diff
@@ -447,6 +447,43 @@ describe(getName(), () => {
       expect(res.releases).toHaveLength(1);
       expect(httpMock.getTrace()).toMatchSnapshot();
     });
+    it('uses quay api', async () => {
+      const tags = [{ name: '1.0.0' }];
```
Comment (viceice) [inline lib/datasource/docker/index.spec.ts:451]:
```suggestion
      const tags = [{ name: '5.0.12' }];
```

File: lib/datasource/docker/index.spec.ts [hunk 13]
```diff
@@ -447,6 +447,43 @@ describe(getName(), () => {
...
+      httpMock
+        .scope('https://quay.io')
+        .get(
+          '/api/v1/repository/bitnami/redis/tag/?limit=100&page=1&onlyActiveTags=true'
+        )
+        .reply(200, { tags, has_additional: false })
+        .get('/v2/')
+        .reply(200, '', {})
+        .get('/v2/bitnami/redis/manifests/1.0.0')
```
Comment (viceice) [inline lib/datasource/docker/index.spec.ts:460]:
```suggestion
        .get('/v2/bitnami/redis/manifests/5.0.12')
```

**Review by viceice — COMMENTED** [review]
just a last small change, you can directly apply suggestions 🙃

**Review by viceice — APPROVED** [review]

**Pull request merged**